### PR TITLE
[ao][sparsity] Implementation of register_layer() and get_mask()

### PR DESCRIPTION
**Pull request by macandro96**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81814
* #80890
* #80889
* #80888
* __->__ #80887
* #80886

The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

### CODE REVIEW

File: test/ao/sparsity/test_activation_sparsifier.py [hunk 1]
```diff
@@ -57,6 +60,48 @@ def reduce_fn(x):
...
+            'mask_fn': _vanilla_norm_sparsifier
+        }
+        sparse_config_layer1 = {'sparsity_level': 0.3}
+
+        register_layer2_args = {
+            'layer': model.linear1,
+            'features': [0, 10, 234],
+            'feature_dim': 1,
+            'mask_fn': _vanilla_norm_sparsifier
+        }
```
Comment (z-a-f) [inline test/ao/sparsity/test_activation_sparsifier.py:99]: If these are hardcoded, I am not sure we need a dedicated getter method for it? Can't we just use `_vanilla...` without the wrapping method?

File: test/ao/sparsity/test_activation_sparsifier.py [hunk 2]
```diff
@@ -78,6 +123,79 @@ def check_constructor(self, activation_sparsifier, model, defaults, sparse_confi
             # all the keys in combined_defaults should be present in sparsifier defaults
             assert config == combined_defaults.get(key, None)
 
+    def check_register_layer(self, activation_sparsifier, defaults, sparse_config, layer_args_list):
```
Comment (z-a-f) [inline test/ao/sparsity/test_activation_sparsifier.py:126]: Again, it looks like the checks are within a generic testcase class, but they are only applicable to your test class. Would be worthhile merging them

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py [hunk 3]
```diff
@@ -95,24 +99,104 @@ def __init__(self, model: nn.Module, aggregate_fn=None, reduce_fn=None, mask_fn=
 
         self.state: Dict[str, Any] = defaultdict(dict)  # layer name -> mask
 
+    @staticmethod
+    def __safe_rail_checks(args):
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:103]: No "private" classes -- we discourage the use of name mangling. Please, use a single underscore

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py [hunk 4]
```diff
@@ -95,24 +99,104 @@ def __init__(self, model: nn.Module, aggregate_fn=None, reduce_fn=None, mask_fn=
...
-
         Note::
-            There is no need to pass in the name of the layer as it is automatically computed as per
-            the fqn convention.
+            - There is no need to pass in the name of the layer as it is automatically computed as per
+              the fqn convention.
 
-            All the functions (fn) passed as argument will be called at a dim, feature level.
+            - All the functions (fn) passed as argument will be called at a dim, feature level.
         """
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:171]: Can we move these changes into the previous PR?

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py [hunk 5]
```diff
@@ -95,24 +99,104 @@ def __init__(self, model: nn.Module, aggregate_fn=None, reduce_fn=None, mask_fn=
...
+    def __safe_rail_checks(args):
+        """Makes sure that some of the functions and attributes are not passed incorrectly
+        """
+
+        # if features are not None, then feature_dim must not be None
+        features, feature_dim = args['features'], args['feature_dim']
+        if features is not None:
+            assert feature_dim is not None, "need feature dim to select features"
+
+        # all the *_fns should be a function
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:112]: nit: "...should be callable"

**Review by z-a-f — CHANGES_REQUESTED** [review]

File: test/ao/sparsity/test_activation_sparsifier.py [hunk 6]
```diff
@@ -94,3 +212,25 @@ def test_activation_sparsifier(self):
...
+        # STEP 2: Register some layers
+        layer_args_list = _ActivationSparsifierUtils._get_register_layer_args(model)
+
+        # Registering..
+        for layer_args in layer_args_list:
+            layer_arg, sparse_config_layer = layer_args
+            activation_sparsifier.register_layer(**layer_arg, **sparse_config_layer)
+
+        # check if things are registered correctly
+        self.check_register_layer(activation_sparsifier, defaults, sparse_config, layer_args_list)
```
Comment (jerryzh168) [inline test/ao/sparsity/test_activation_sparsifier.py:225]: also I'm wondering if we can separate these steps into multiple tests to keep the test itself shorter

File: test/ao/sparsity/test_activation_sparsifier.py [hunk 7]
```diff
@@ -78,6 +123,79 @@ def check_constructor(self, activation_sparsifier, model, defaults, sparse_confi
...
+        assert len(data_groups) == len(layer_args_list)
+        for layer_args in layer_args_list:
+            layer_arg, sparse_config_layer = layer_args
+
+            # check sparse config
+            sparse_config_actual = copy.deepcopy(sparse_config)
+            sparse_config_actual.update(sparse_config_layer)
+
+            name = module_to_fqn(activation_sparsifier.model, layer_arg['layer'])
+            assert data_groups[name]['layer'] == layer_arg['layer']
```
Comment (jerryzh168) [inline test/ao/sparsity/test_activation_sparsifier.py:155]: also I think ideally we should focus on testing the functionality instead of implementation details, it looks like we are testing implementation details here

File: test/ao/sparsity/test_activation_sparsifier.py (hunk 1 above)
Comment (macandro96) [reply test/ao/sparsity/test_activation_sparsifier.py]: done

File: test/ao/sparsity/test_activation_sparsifier.py (hunk 6 above)
Comment (macandro96) [reply test/ao/sparsity/test_activation_sparsifier.py]: Since we're testing the complete workflow here, the state of sparsifier will change at each step. So, we have one single test-case here.

File: test/ao/sparsity/test_activation_sparsifier.py (hunk 7 above)
Comment (macandro96) [reply test/ao/sparsity/test_activation_sparsifier.py:86]: removed such tests

File: test/ao/sparsity/test_activation_sparsifier.py (hunk 2 above)
Comment (macandro96) [reply test/ao/sparsity/test_activation_sparsifier.py]: done

**Review by z-a-f — APPROVED** [review]
